File: Backend/models/role_application.go
package models

import (
	"encoding/json"
	"time"

	"gorm.io/gorm"
)

// RoleType represents the type of role application
type RoleType string

const (
	RoleTypeWorker RoleType = "worker"
	RoleTypeBroker RoleType = "broker"
)

// ApplicationStatus represents the status of role application
type ApplicationStatus string

const (
	ApplicationStatusPending  ApplicationStatus = "pending"
	ApplicationStatusApproved ApplicationStatus = "approved"
	ApplicationStatusRejected ApplicationStatus = "rejected"
)

// Enhanced JSON types for better frontend consumption
type ContactInfo struct {
	AlternativeNumber string `json:"alternative_number"`
}

type WorkerAddress struct {
	Street   string `json:"street"`
	City     string `json:"city"`
	State    string `json:"state"`
	Pincode  string `json:"pincode"`
	Landmark string `json:"landmark"`
}

type BankingInfo struct {
	AccountNumber     string `json:"account_number"`
	IfscCode         string `json:"ifsc_code"`
	BankName         string `json:"bank_name"`
	AccountHolderName string `json:"account_holder_name"`
}

type Documents struct {
	AadharCard        string `json:"aadhar_card"`
	PanCard          string `json:"pan_card"`
	ProfilePic       string `json:"profile_pic"`
	PoliceVerification string `json:"police_verification,omitempty"`
}

// Enhanced Worker with parsed JSON
type EnhancedWorker struct {
	gorm.Model
	UserID             uint       `json:"user_id"`
	RoleApplicationID  *uint      `json:"role_application_id"`
	WorkerType         WorkerType `json:"worker_type"`
	
	// Parsed JSON fields
	ContactInfo        ContactInfo    `json:"contact_info"`
	Address            WorkerAddress  `json:"address"`
	BankingInfo        BankingInfo    `json:"banking_info"`
	Documents          Documents   `json:"documents"`
	Skills             []string    `json:"skills"`
	
	// Regular fields
	Experience         int        `json:"experience_years"`
	IsAvailable        bool       `json:"is_available"`
	Rating             float64    `json:"rating"`
	TotalBookings      int        `json:"total_bookings"`
	Earnings           float64    `json:"earnings"`
	TotalJobs          int        `json:"total_jobs"`
	IsActive           bool       `json:"is_active"`
}

// Enhanced Broker with parsed JSON
type EnhancedBroker struct {
	gorm.Model
	UserID             uint    `json:"user_id"`
	RoleApplicationID  *uint   `json:"role_application_id"`
	
	// Parsed JSON fields
	ContactInfo        ContactInfo    `json:"contact_info"`
	Address            WorkerAddress  `json:"address"`
	Documents          Documents      `json:"documents"`
	
	// Broker specific
	License            string  `json:"license"`
	Agency             string  `json:"agency"`
	IsActive           bool    `json:"is_active"`
}

// Enhanced RoleApplication with parsed JSON data
type EnhancedRoleApplication struct {
	gorm.Model
	UserID           uint             `json:"user_id"`
	RequestedRole    string           `json:"requested_role"`
	Status           ApplicationStatus `json:"status"`
	SubmittedAt      time.Time        `json:"submitted_at"`
	ReviewedAt       *time.Time       `json:"reviewed_at"`
	ReviewedBy       *uint            `json:"reviewed_by"`
	
	// Relationships
	User             User             `json:"user"`
	ReviewedByUser   *User            `json:"reviewed_by_user"`
	Worker           *EnhancedWorker  `json:"worker,omitempty"`
	Broker           *EnhancedBroker  `json:"broker,omitempty"`
}

// RoleApplication represents the simplified role application model
type RoleApplication struct {
	gorm.Model
	UserID           uint             `json:"user_id" gorm:"not null"`
	RequestedRole    string           `json:"requested_role" gorm:"not null"`
	Status           ApplicationStatus `json:"status" gorm:"default:'pending'"`
	SubmittedAt      time.Time        `json:"submitted_at" gorm:"default:NOW()"`
	ReviewedAt       *time.Time       `json:"reviewed_at"`
	ReviewedBy       *uint            `json:"reviewed_by"`
	
	// Relationships
	User             User             `json:"user" gorm:"foreignKey:UserID"`
	ReviewedByUser   *User            `json:"reviewed_by_user" gorm:"foreignKey:ReviewedBy"`
	Worker           *Worker          `json:"worker,omitempty" gorm:"foreignKey:RoleApplicationID"`
	Broker           *Broker          `json:"broker,omitempty" gorm:"foreignKey:RoleApplicationID"`
}

// TableName returns the table name for RoleApplication
func (RoleApplication) TableName() string {
	return "role_applications"
}

// ConvertToEnhanced converts a RoleApplication to EnhancedRoleApplication with parsed JSON
func (ra *RoleApplication) ConvertToEnhanced() *EnhancedRoleApplication {
	enhanced := &EnhancedRoleApplication{
		Model:         ra.Model,
		UserID:        ra.UserID,
		RequestedRole: ra.RequestedRole,
		Status:        ra.Status,
		SubmittedAt:   ra.SubmittedAt,
		ReviewedAt:    ra.ReviewedAt,
		ReviewedBy:    ra.ReviewedBy,
		User:          ra.User,
		ReviewedByUser: ra.ReviewedByUser,
	}

	// Convert Worker data if exists
	if ra.Worker != nil {
		enhancedWorker := &EnhancedWorker{
			Model:            ra.Worker.Model,
			UserID:           ra.Worker.UserID,
			RoleApplicationID: ra.Worker.RoleApplicationID,
			WorkerType:       ra.Worker.WorkerType,
			Experience:       ra.Worker.Experience,
			IsAvailable:      ra.Worker.IsAvailable,
			Rating:           ra.Worker.Rating,
			TotalBookings:    ra.Worker.TotalBookings,
			Earnings:         ra.Worker.Earnings,
			TotalJobs:        ra.Worker.TotalJobs,
			IsActive:         ra.Worker.IsActive,
		}

		// Parse JSON fields
		if ra.Worker.ContactInfo != "" {
			json.Unmarshal([]byte(ra.Worker.ContactInfo), &enhancedWorker.ContactInfo)
		}
		if ra.Worker.Address != "" {
			json.Unmarshal([]byte(ra.Worker.Address), &enhancedWorker.Address)
		}
		if ra.Worker.BankingInfo != "" {
			json.Unmarshal([]byte(ra.Worker.BankingInfo), &enhancedWorker.BankingInfo)
		}
		if ra.Worker.Documents != "" {
			json.Unmarshal([]byte(ra.Worker.Documents), &enhancedWorker.Documents)
		}
		if ra.Worker.Skills != "" {
			json.Unmarshal([]byte(ra.Worker.Skills), &enhancedWorker.Skills)
		}

		enhanced.Worker = enhancedWorker
	}

	// Convert Broker data if exists
	if ra.Broker != nil {
		enhancedBroker := &EnhancedBroker{
			Model:            ra.Broker.Model,
			UserID:           ra.Broker.UserID,
			RoleApplicationID: ra.Broker.RoleApplicationID,
			License:          ra.Broker.License,
			Agency:           ra.Broker.Agency,
			IsActive:         ra.Broker.IsActive,
		}

		// Parse JSON fields
		if ra.Broker.ContactInfo != "" {
			json.Unmarshal([]byte(ra.Broker.ContactInfo), &enhancedBroker.ContactInfo)
		}
		if ra.Broker.Address != "" {
			json.Unmarshal([]byte(ra.Broker.Address), &enhancedBroker.Address)
		}
		if ra.Broker.Documents != "" {
			json.Unmarshal([]byte(ra.Broker.Documents), &enhancedBroker.Documents)
		}

		enhanced.Broker = enhancedBroker
	}

	return enhanced
}

File: Admin/src/types/roleApplication.ts
export interface User {
  ID: number;
  id?: number; // For compatibility with existing components
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;
  name: string;
  email?: string | null;
  phone: string;
  user_type: string;
  avatar: string;
  gender: string;
  is_active: boolean;
  last_login_at?: string | null;
  role_application_status: string;
  application_date?: string | null;
  approval_date?: string | null;
  wallet_balance: number;
  subscription_id?: number | null;
  subscription?: unknown | null;
  has_active_subscription: boolean;
  subscription_expiry_date?: string | null;
}

// Enhanced JSON types for better frontend consumption
export interface ContactInfo {
  alternative_number: string;
}

export interface WorkerAddress {
  street: string;
  city: string;
  state: string;
  pincode: string;
  landmark: string;
}

export interface BankingInfo {
  account_number: string;
  ifsc_code: string;
  bank_name: string;
  account_holder_name: string;
}

export interface Documents {
  aadhar_card: string;
  pan_card: string;
  profile_pic: string;
  police_verification?: string;
}

// Enhanced Worker with parsed JSON
export interface EnhancedWorker {
  id: number;
  user_id: number;
  role_application_id?: number;
  worker_type: string;

  // Parsed JSON fields
  contact_info: ContactInfo;
  address: WorkerAddress;
  banking_info: BankingInfo;
  documents: Documents;
  skills: string[];

  // Regular fields
  experience_years: number;
  is_available: boolean;
  rating: number;
  total_bookings: number;
  earnings: number;
  total_jobs: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// Enhanced Broker with parsed JSON
export interface EnhancedBroker {
  id: number;
  user_id: number;
  role_application_id?: number;

  // Parsed JSON fields
  contact_info: ContactInfo;
  address: WorkerAddress;
  documents: Documents;

  // Broker specific
  license: string;
  agency: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// Enhanced RoleApplication with parsed JSON data
export interface EnhancedRoleApplication {
  ID: number;
  id?: number; // For compatibility with existing components
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;
  user_id: number;
  requested_role: string;
  status: string;
  submitted_at: string;
  reviewed_at?: string | null;
  reviewed_by?: number | null;

  // Relationships
  user: User;
  reviewed_by_user?: User | null;
  worker?: EnhancedWorker;
  broker?: EnhancedBroker;
}

// Legacy interface for backward compatibility
export interface RoleApplication {
  id: number;
  user_id: number;
  requested_role: string;
  status: string;
  admin_notes?: string;
  submitted_at: string;
  reviewed_at?: string;
  reviewed_by?: number;

  // Worker-specific fields
  worker_type?: string;
  skills?: string[];
  experience_years?: number;
  bank_details?: {
    account_holder: string;
    account_number: string;
    ifsc_code: string;
    bank_name: string;
    bank_branch: string;
  };

  // Broker-specific fields
  broker_license?: string;
  broker_agency?: string;

  // Common fields
  address: {
    city: string;
    state: string;
    country: string;
    address: string;
    postal_code: string;
    latitude?: number;
    longitude?: number;
  };
  documents: {
    aadhaar_card: string;
    pan_card: string;
    police_verification?: string;
    profile_pic: string;
  };

  created_at: string;
  updated_at: string;

  // Relationships
  user?: User;
}

export interface RoleApplicationFilterState {
  search: string;
  status: string;
  requested_role: string;
  date_from: string;
  date_to: string;
}

export interface RoleApplicationsApiResponse {
  applications: EnhancedRoleApplication[];
  pagination: {
    current_page: number;
    total_pages: number;
    total: number;
    limit: number;
  };
}

export interface BackendApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
  timestamp: string;
}
